Content import composability refactor and single resource import

rtibbles opened the pull request:
## Summary
* Adds the ability to import content metadata from JSON rather than a SQLite database
* Updates SQLAlchemy version to 1.4 and makes syntax updates
* Allows for import of partial channel data (sub trees rather than the whole tree) and incremental import of the entire tree
* Creates a public API endpoint to return the required metadata JSON for importing a single resource
* Does some refactors to make content import functionality more composable
* Creates an async task for importing a single resource over the network

## References
Fixes #9552
Fixes #9492
Fixes #9491
Completes some of #9266
Last commit also fixes https://github.com/learningequality/kolibri-app/issues/105, which might be cherry-picked to https://github.com/learningequality/kolibri/pull/9683 if that is approved for merge first.

## Reviewer guidance
Do the tests of the new import functionality make sense? Does importing a resource from another instance work?

Does the way I've split up the different functionality of the import manager classes feel readable/practical/sustainable?

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Update sqlalchemy.

rtibbles committed:
Update sqlalchemy selects to use new arg syntax.

rtibbles committed:
Add support for partial imports of channel metadata.

rtibbles committed:
Handle in memory databases better, and don't try to attach them.

rtibbles committed:
Clean up unused functions and give better errors.

rtibbles committed:
Create public import metadata viewset.

rtibbles committed:
Migrate progress tracker functionality to mixin.
Add file descriptor safe executor factory function.

rtibbles committed:
Add async task for importing a specific resource.

rtibbles committed:
Always merge models in postgres for partial imports.

rtibbles commented:
DMG build failure appears to be unrelated to these changes. Tracked separately: https://github.com/learningequality/kolibri-app/issues/105

vkWeb commented on kolibri/core/tasks/utils.py:
```diff
@@ -206,3 +212,142 @@ def db_connection():
+                self.job.progress + increment, self.job.total_progress
+            )
+            if extra_data and isinstance(extra_data, dict):
+                self.job.update_job_metadata(**extra_data)
```
Umm, I don't see a method `update_job_metadata` defined on `Job` class. Shouldn't we be calling `self.update_job_metadata(**extra_data)` here?

rtibbles replied:
Ah, good catch - I think I meant to add it to the Job class, because it seemed like a useful thing to be able to set directly on the job, but then didn't...

vkWeb replied:
Keeping that function at Job class makes sense to me sir

vkWeb commented on kolibri/core/tasks/utils.py:
```diff
@@ -206,3 +212,142 @@ def db_connection():
+            return True
+
+    def cancel(self):
+        return self.check_for_cancel()
```
Shouldn't we be cancelling the job here? umm, am I missing something sir?

rtibbles replied:
This is copied from the original implementation, so I haven't actually changed anything here - but you're right the method name is misleading, what it's doing is reraising a handled cancel, rather than actually initiating a cancel.

vkWeb replied:
Yeap sir then we should cancel the job there.

rtibbles replied:
I think it's actually better to rename the method, as it is never used to initiate a cancel from within the task, only to reraise the cancellation once the task has finished its cleanup tasks.

vkWeb reviewed (commented):
@rtibbles importing from another kolibri instance is working as expected on my manual test. 

Though the progress update percentage is being reported wrong: 

![Screenshot from 2022-10-03 18-36-49](https://user-images.githubusercontent.com/26724128/193588011-115c01c1-dbb8-4944-89a8-0caeeb6bb6b2.png)

rtibbles commented:
That's quite a high percentage!

vkWeb commented:
> That's quite a high percentage!

Very high indeed sir!

bjester commented on kolibri/core/content/public_api.py:
```diff
@@ -0,0 +1,105 @@
+            from_contentnode_id__in=node_ids, to_contentnode_id__in=node_ids
+        )
+        channel_metadata = models.ChannelMetadata.objects.filter(id=node.channel_id)
+
```
I doubt this would occur with a list of ancestor nodes, but do any of these `__in` filters in this method above need chunking for SQLite? Or is that handled at a global level-- I can't recall.

rtibbles replied:
I think as `node_ids` has not been coerced to a list it should still be being passed as a subquery rather than as an explicit list that would need SQL variables for interpolation, so I think this is safe.

bjester commented on kolibri/core/content/public_api.py:
```diff
@@ -0,0 +1,105 @@
+
+        data["schema_version"] = content_schema
+
+        return Response(data)
```
So we can mimic this endpoint on Studio to allow remote browsing?

rtibbles replied:
This endpoint is only for import metadata, getting all relevant metadata for importing a single resource (rather than the entire channel database). There are some already implemented public API endpoints that would be needed in addition (a V2 public channel endpoint and several content node endpoints).

They are all exposed under a v2 content metadata router: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/public/api_urls.py#L47

bjester commented on kolibri/core/content/utils/channel_import.py:
```diff
@@ -652,73 +702,40 @@ def can_use_sqlite_attach_method(self, model, table_mapper):
+                    return False
+                return True
+            elif existing_channel["version"] < self.channel_version:
                 # We have an older version of this channel, so let's clean out the old stuff first
```
Seems this should inherently and correctly handle the case where the previous import was a partial import?

rtibbles replied:
As it just deletes everything, it will handle the case when the previous import was partial or when it was complete. We add the check for a current partial import, because we can only do that in the case where the currently installed version of the channel is the same, otherwise we run the risk of mismatched/colliding tree metadata.

bjester reviewed (commented):
Code review looks good! 

In testing, I got an error for a channel in which a new version was available (below), and I also saw the progress % issue @vkWeb noted. 

I clicked `View Changes` and got this
```
ERROR    Job 6c501be7ac2d4a25a2322a2b631858c7 raised an exception: Traceback (most recent call last):
  File "/kolibri/core/tasks/job.py", line 261, in execute
    result = func(*args, **kwargs)
  File "/kolibri/core/tasks/registry.py", line 221, in __call__
    return self.func(*args, **kwargs)
  File "/kolibri/core/content/tasks.py", line 500, in remotechanneldiffstats
    baseurl=baseurl,
  File "/kolibri/core/content/utils/upgrade.py", line 79, in diff_stats
    destination=destination_path,
  File "/kolibri/core/content/utils/channel_import.py", line 1150, in initialize_import_manager
    min_version = channel_metadata.get(
AttributeError: 'unicode' object has no attribute 'get'
```

rtibbles commented:
Looks like I've messed up the argument handling for the diff file stats too, will fix!

rtibbles committed:
Move checksum checking into file transfer logic.

rtibbles committed:
Simplify job cancellation handling and metadata saving.

rtibbles committed:
Update upgrade usage of initialize_import_manager to new function signature.

rtibbles committed:
Set the total job progress when doing start progress.

rtibbles commented:
Have fixed the import diff stats issue, upgrading now works properly.

I did see a frontend issue during the upgrade process, but the backend experienced no errors. Not sure if the issue is also extant on develop, as I've not touched the frontend here.

Progress updating is now working properly again - I had simply forgotten to set the total progress at the start!
![image](https://user-images.githubusercontent.com/1680573/201250942-5e1983c7-b35b-4625-8fcc-0b9d7b75be19.png)

bjester commented:
Upgrading a channel, I got the following frontend error:
```
vue.runtime.esm.js:1897 TypeError: Cannot read properties of undefined (reading 'id')
    at NewChannelVersionPage.vue:237:1
logError @ vue.runtime.esm.js:1897
```

Also, does this mean a resource was a duplicate or that it failed to import?
![Screenshot from 2022-12-05 11-52-39](https://user-images.githubusercontent.com/819838/205731255-b1c75434-8a86-4cde-8db3-eb2775a37e57.png)

bjester commented on kolibri/core/content/utils/resource_import.py:
```diff
@@ -0,0 +1,513 @@
+                exclude_node_ids = None
+            except ContentManifestParseError:
+                pass
+            manifest_file.close()
```
I think this manifest parsing would be better placed in a factory function that builds a class instance from it, instead of having it in the class constructor. The reason is that it takes precedence on determining the values of `node_ids` and `exclude_node_ids`, but if it were to fail, it makes that codepath a bit ambiguous on the intended outcome since it just suppresses the parse error.

I see an inherited class is also utilizing this argument, so a static class method might be best for the factory. I would expect that the factory function wouldn't accept `node_ids` and `exclude_node_ids` arguments, thereby making the API clear that it's one or the other.

rtibbles commented:
> Upgrading a channel, I got the following frontend error:

I got the same issue here, it seemed like a frontend only issue, but can double check this and fix (I think it might be pre-existing in develop but worth fixing here).

> Also, does this mean a resource was a duplicate or that it failed to import?

Yeah, this is a pre-existing issue with resource counting being deduped in one place and not in another. I'll fix it in this PR in case it doesn't get fixed in 0.15.

rtibbles committed:
Cleanup and first pass implementation of non-SQL based metadata import.

rtibbles committed:
Add factory functions for manifest file handling.

rtibbles committed:
Fix refactor induced bugs in channel update workflow.

rtibbles commented:
Updated - add a `from_manifest` class method to handle the manifest construction and make the keyword arguments explicit.

Cleaned up various refactor induced bugs in the channel update workflow. Should be good to go!

rtibbles committed:
Migrate resource import logic out of management command to make it more composable.

bjester reviewed (approved):
LGTM! :100:
